src/global-firestore-data/courses.ts

```diff
@@ -0,0 +1,51 @@
+import { doc, getDoc } from 'firebase/firestore';
+import { coursesCollection, availableRostersForCoursesCollection } from '../firebase-config';
+
+/**
+ * This function uses the subject and number of a course to retrieve the course information from 'courses' collection
+ * @param season The season that the course is under in the user's semester view
+ * @param year The year the course is under in the user's semester view
+ * @param subject The subject of the course
+ * @param number The number of the course
+ * @returns `Promise<CornellCourseRosterCourseFullDetail>`
+ */
+export const getCourseWithSeasonAndYear = async (
+  season: FirestoreSemesterSeason,
+  year: number,
+  subject: string,
+  number: string
+): Promise<CornellCourseRosterCourseFullDetail> => {
+  const roster = seasonAndYearToRosterIdentifier(season, year);
+  const course = await getDoc(doc(coursesCollection, `${roster}/${subject}/${number}`));
+  if (!course.exists()) {
+    return getLastOffering(subject, number);
+  }
+  return course.data()?.course as CornellCourseRosterCourseFullDetail;
```

Comment:
Is this cast needed after annotating the return type?

Reply:
While we annotated the return type of the function, we never actually asserted/checked if the object being returned actually matches that type.  In fact, Typescript thinks the type is `any`, which makes sense because it can't possibly infer the type of what the Firestore is returning. 

<img width="915" alt="Screen Shot 2023-03-25 at 11 27 11 PM" src="https://user-images.githubusercontent.com/46412997/227753594-b5346047-0f10-4056-b580-0b0d3f1d03ae.png">

So I just like to use type assertion before returning as a sanity check.

Reply:
Does removing the cast produce a type error? Casts in typescript don't actually perform any runtime validation.

Reply:
It doesn't produce a type error, but is specifying the return type of the function enough? `course.data()?.course` is of type `DocumentData`, but our function returns `CornellCourseRosterCourseFullDetail`. This means we are just hoping that `DocumentData === CornellCourseRosterCourseFullDetail`, without checking at all.

Reply:
Yeah like what I'm saying is that typescript casts are just straight up removed during compilation to javascript, so even if `DocumentData === CornellCourseRosterCourseFullDetail`, the cast wouldn't fail.

Reply:
Gotcha, then I will just define `course` with type `CornellCourseRosterCourseFullDetail`

Reply:
Wait I can't do that either, nvm ill just nuke the type cast